oscarapi/middleware.py

```diff
@@ -0,0 +1,167 @@
+    >>>
+    >>> request.META['HTTP_SESSION_ID'] = 'SID:BULLSHIT:example.com:987171879'
+    >>> parse_session_id(request)
+    
+    >>> request.META['HTTP_SESSION_ID'] = 'ENTIREGABRBAGE'
+    >>> parse_session_id(request)
+    
+    >>> request.META['HTTP_SESSION_ID'] = 'SID:ANON:987171879'
+    >>> parse_session_id(request)
+    
+    """
+    unparsed_session_id = request.META.get('HTTP_SESSION_ID', None)
+    if unparsed_session_id is not None:
+        parsed_session_id = HTTP_SESSION_ID_REGEX.match(unparsed_session_id)
+        if parsed_session_id is not None:
+            return parsed_session_id.groupdict()
+
+    return None
+
+
+def start_or_resume(session_id, session_type):
+    if session_type == 'ANON':
+        return get_session(session_id, raise_on_create=False)
+
+    return get_session(session_id, raise_on_create=True)
+
+
+def is_api_request(request):
+    path = request.path.lower()
+    api_root = reverse('api-root').lower()
+    return path.startswith(api_root)
+
+
+class HeaderSessionMiddleware(SessionMiddleware):
+    """
+    Implement session through headers:
+
+    http://www.w3.org/TR/WD-session-id
+
+    TODO:
+    Implement gateway protection, with permission options for usage of
+    header sessions. With that in place the api can be used for both trusted
+    and non trusted clients, see README.rst.
+    """
+
+    def process_request(self, request):
+        """
+        Parse the session id from the 'Session-Id: ' header when using the api.
+        """
+        if is_api_request(request):
+            try:
+                parsed_session_uri = parse_session_id(request)
+                if parsed_session_uri is not None:
+                    domain = get_domain(request)
+                    if parsed_session_uri['realm'] != domain:
+                        raise exceptions.NotAcceptable(
+                            _('Can not accept cookie with realm %s on realm %s') % (
+                                parsed_session_uri['realm'],
+                                domain
+                            )
+                        )
+                    session_id = session_id_from_parsed_session_uri(
+                        parsed_session_uri)
+                    request.session = start_or_resume(
+                        session_id, session_type=parsed_session_uri['type'])
+                    request.parsed_session_uri = parsed_session_uri
+
+                    # since the session id is assigned by the CLIENT, there is
+                    # no point in having csrf_protection. Session id's read
+                    # from cookies, still need csrf!
+                    request.csrf_processing_done = True
+                    return None
+            except exceptions.APIException as e:
+                response = HttpResponse('{"reason": "%s"}' % e.detail)
+                response.status_code = e.status_code
+                return response
```

Comment:
Not necessarily, I might want to use a rest_framework.response.Response instead of HttpResponse. I'll have to test that to see if the object get's properly unwrapped into a HttpResponse if thrown somewhere in a middleware. If that doesn't work, you've got a point.